Aeson-1.5 and QuickCheck-2.14: revbumps needed

ezzieyguywuf opened the pull request:
I am splitting these out of #1118 in order to make #1118 smaller and more
managenable.

There are various conflicts that need to be fixed before all of these versions
can be used, and that will become the focus of #1118. In the meantime, anything
that causes an issue (that is added by this PR) has been masked.

trofi commented:
I geat at least 2 conflicts on it:

```
$ emerge -pve1 @all-haskell --ignore-built-slot-operator-deps=y --verbose-conflicts
...
  (dev-haskell/unordered-containers-0.2.13.0:0/0.2.13.0::haskell, ebuild scheduled for merge) USE="doc hscolour profile test -debug -hoogle" ABI_X86="(64)" conflicts with
    <dev-haskell/unordered-containers-0.2.13 required by (dev-haskell/req-3.6.0:0/3.6.0::haskell, ebuild scheduled for merge) USE="doc hscolour profile test -hoogle" ABI_X86="(64)"
    ^                                 ^^^^^^

dev-haskell/github:0

  (dev-haskell/github-0.26:0/0.26::haskell, ebuild scheduled for merge) USE="doc hscolour openssl profile test -hoogle" ABI_X86="(64)" conflicts with
    <dev-haskell/github-0.24:= required by (dev-haskell/github-backup-1.20200721:0/0::haskell, ebuild scheduled for merge) USE="" ABI_X86="(64)"
    ^           
```

trofi commented on dev-haskell/uuid-types/uuid-types-1.0.3-r1.ebuild:
```diff
@@ -18,10 +18,10 @@ KEYWORDS="~amd64 ~x86"
 IUSE=""
 
 RDEPEND=">=dev-haskell/binary-0.4:=[profile?] <dev-haskell/binary-0.9:=[profile?]
-	>=dev-haskell/random-1.0.1:=[profile?] <dev-haskell/random-1.3:=[profile?]
```
When `RDEPEND` changes we usually have to bump ebuild's revision (from `1.0.3-r1` to `1.0.3-r2`) to force ebuild upgrade. Otherwise `portage` keeps to locally installed ebuild with it's upper bounds and produces very confusing errors on upgrade.

ezzieyguywuf replied:
Yes I'm aware of this and have been trying to do the rename as you've listed. If I've missed some it shouldn't be a problem at all to tweak it.

ezzieyguywuf commented:
> I geat at least 2 conflicts on it:

Dang, not sure how mine missed those.

I'll dig into it.

trofi commented on dev-haskell/svg-builder/svg-builder-0.1.1-r1.ebuild:
```diff
@@ -0,0 +1,29 @@
+# Copyright 1999-2021 Gentoo Authors
```
Usually when you do dependency-loosening version bumps I suggest just renaming ebuilds instead of copying those. That way it's easier to review actual changes (and chance of a revert is quite small).

ezzieyguywuf replied:
>  I suggest just renaming ebuilds instead of copying those

I'm afraid I don't understand. The original ebuild was `svg-builder-0.1.1.ebuild` and I named the new one `svg-builder-0.1.1-r1.ebuild` due to the `RDEPEND` changes. Specifically, this is the process I followed:

```sh
cd ${gentoo_haskell}/dev-haskell/svg-builder
hackport merge svg-builder # pulls in the new upperbounds from hackage
git diff # what changed? since rdeps changed...

mv svg-builder-0.1.1.ebuild svg-builder-0.1.1-r1.ebuild
# need to recover the original file
git checkout HEAD~ -- svg-builder-0.1.1.ebuild

# finally, all the end-matter
pkgcheck scan # any issues? fix and commit
repoman full -dx # same
git add .
git commit -s -m "dev-haskell/.....you-get-the-point" # I know I could use repoman commit here
```

Can you elaborate a bit on what you mean by "just renaming ebuilds"?

trofi replied:
`git checkout HEAD~ -- svg-builder-0.1.1.ebuild` restores original file. I think we can safely remove it as part of this change.

ezzieyguywuf replied:
Ah hah, ok I will remove the files whenever I do the `-rN` bump

trofi commented on dev-haskell/svg-builder/svg-builder-0.1.1-r1.ebuild:
```diff
@@ -0,0 +1,29 @@
+"
+DEPEND="${RDEPEND}
+	>=dev-haskell/cabal-1.16.0
+"
```
Your comment says you used hackage's revision. But this change does not change `.cabal`'s file. I would expect it to hold outdated version bounds.

ezzieyguywuf replied:
The change is seen as a revision in hackage, the package was bumped to "r2" (we missed "r1").

Thus, hackport (in all it's infinite awesomeness) recognizes this and generates the new ebuild, despite the version number being the same.

trofi replied:
Package revisions on hackage are done in a way that change `.cabal` filed in index file stored at `01-index.tar`. They don't update `.tar.gz` file of a released package. Currently hackport does not download revised `.cabal` file separately. It should.

ezzieyguywuf replied:
HAH, ok I finally get it.

```sh
ebuild ${haskell_repo}/dev-haskell/svg-builder-0.1.1-r1.ebuild prepare
cat /var/tmp/portage/dev-haskell/svg-builder-0.1.1-r1/work/svg-builder-0.1.1-r1/svg-builder.cabal | grep `base`
```

This shows what you're saying, that despite the updates on hackages *index*, the actual tarball hasn't been changed. Thus we still need the `cabal_chdeps` stuff.

I think I knew this, but for some reason I wasn't quite comprehending what you were saying. I can fix that in this commit and others, no problem.

trofi reviewed (changes requested):
I also added two comments that probably apply to many of your commits. I hope it's not too much hassle to tweak them.

ezzieyguywuf commented:
> I geat at least 2 conflicts on it:

Ok, rebased on latest master, resolved the `github-backup` conflict. I was unable to reproduce `unordered-containers`, I explicitly got `unordered-containers-0.3.0.1` and `req-0.3.6.0` in my update list with no issues.

Maybe try `--backtrack=100`?

trofi commented on dev-haskell/nats/nats-1.1.2-r1.ebuild:
```diff
@@ -20,14 +20,18 @@ SLOT="0/${PV}"
 KEYWORDS="~amd64 ~x86"
 IUSE="+binary +hashable +template-haskell"
 
-RESTRICT=test # avoid circular dep: nats[hashable]->hashable[test]->quickcheck->nats
```
Was removal of `RESTRICT` intentional?

ezzieyguywuf replied:
Yes I believe so, however the test cannot be run until the newer aeson is unmasked, so I'm going to pop this commit out of this PR and put it back into #1118

ezzieyguywuf commented:
@trofi found a typo in one of the ebuilds. I added it as a single commit for now, in case you were in the middle of reviewing the current branch.

I can merge the fix into the revbump commit and force push whenever, just let me know.

trofi commented on dev-haskell/diagrams-lib/diagrams-lib-1.4.3.ebuild:
```diff
@@ -37,7 +39,7 @@ RDEPEND=">=dev-haskell/active-0.2:=[profile?] <dev-haskell/active-0.3:=[profile?
 	>=dev-haskell/linear-1.20.1:=[profile?] <dev-haskell/linear-1.21:=[profile?]
 	>=dev-haskell/monoid-extras-0.3:=[profile?] <dev-haskell/monoid-extras-0.6:=[profile?]
 	>=dev-haskell/mtl-2.0:=[profile?] <dev-haskell/mtl-2.3:=[profile?]
-	>=dev-haskell/optparse-applicative-0.11:=[profile?] <dev-haskell/optparse-applicative-0.16:=[profile?]
```
`RDEPEND` change will need an ebuild revision bump.

trofi commented on dev-haskell/filepath-bytestring/filepath-bytestring-1.4.2.1.6.ebuild:
```diff
@@ -23,3 +23,9 @@ DEPEND="${RDEPEND}
 	>=dev-haskell/cabal-1.18
 	test? ( >=dev-haskell/quickcheck-2.7 <dev-haskell/quickcheck-2.14 )
```
Probably needs a `DEPEND` tweak as well.

trofi reviewed (changes requested):
I pulled a few trivial commits as is (I hope there will be no rebase conflicts) and added a few comments.

ezzieyguywuf commented:
> 
> 
> I pulled a few trivial commits as is (I hope there will be no rebase conflicts) and added a few comments.

@trofi thank you for the review. I rebased with no issues (whoot!) but switching back to draft while I work on your other comments.

ezzieyguywuf commented on dev-haskell/concurrent-supply/concurrent-supply-0.1.8-r1.ebuild:
```diff
@@ -0,0 +1,30 @@
+# Copyright 1999-2021 Gentoo Authors
```
What about fetching the new cabal file from hackage, are you open to that?

trofi replied:
Yeah, having new `.cabal` file in `SRC_URI` is perfectly fine. Just make sure you are linking to an immutable versioned revision of `.cabal` file instead of a live one.

trofi commented on dev-haskell/concurrent-supply/concurrent-supply-0.1.8-r1.ebuild:
```diff
@@ -1,29 +1,37 @@
-SRC_URI="https://hackage.haskell.org/package/${P}/${P}.tar.gz"
+SRC_URI="
+	https://hackage.haskell.org/package/${P}/${P}.tar.gz
+	https://hackage.haskell.org/package/${P}/revision/1.cabal -> ${PN}.cabal"
```
The `distfiles` names have to be unique. Otherwise next bump will reuse (outdated) `${PN}.cabal` file for you on next bump like that. I suggest using something verbose but unambiguous, like `${PF}.cabal`.

The `DIST` entry above is a good test on a filename in `distfiles/`.

Same for other ebuilds.

trofi reviewed (changes requested):
I think I have only one annoying comment :)

trofi reviewed (changes requested):
I also send an invitation to gentoo-haskell team. Once you address it feel free to merge it yourself :)

solpeth commented on dev-haskell/hspec-core/hspec-core-2.7.4.ebuild:
```diff
@@ -0,0 +1,39 @@
+# Copyright 1999-2020 Gentoo Authors
```
Is there a reason for this version as well as `2.7.5`? If not I'll remove it after PR is merged.

ezzieyguywuf replied:
Good catch, I took out `2.7.4`

solpeth reviewed (approved):
Lots to review! I've looked over each file and excluding things already addressed by trofi nothing jumps out at me in particular.

I'm assuming that the manual grabbing of the revised cabal file and inserting it into the build has been cleared as fine (in fact, if this is so I intend to perhaps look into automating this behaviour via the eclass and hackport if it is feasible and wanted).

There are lots of major version bumps (major in the sense of the PVP). I have not checked that the reverse dependencies of these packages remain satisfied, but I have also noticed that a large number of bumped packages will remain masked until later. So I'm assuming that the risk from the size of this PR is outweighed by that masking.

Some packages have had their minor versions bumped twice (e.g. `bifunctors-5.5.8` and `bifunctors-5.5.9`). Unless there is a reason for it (and there are reasons to do this, such as minor version bumps depending on different major versions of dependencies, and masking one until such time as it is ready) I will hopefully have a chance to go through post-merge and clean it all out.

I'm hesitant to merge right now due to my restricted/sporadic availability at the moment, in case there is need for mopping up, but I will try to get to it within the next few days if trofi cannot.

Thank you!

ezzieyguywuf commented:
> I'm assuming that the manual grabbing of the revised cabal file and inserting it into the build has been cleared as fine (in fact, if this is so I intend to perhaps look into automating this behaviour via the eclass and hackport if it is feasible and wanted).

@trofi and I discussed this on IRC - his response was more or less "you can do it if you want."

Personally, I like the fact that it draws a clear distinction between what version bumps have come from upstream versus what we're doing "locally" here in gentoo-haskell.

I'm in favour of automating the change, though I think the appropriate place would be in hackport itself, not in the eclass.



> I have not checked that the reverse dependencies of these packages remain satisfied

I have, and have masked those packages that pose an issue. #1118 will unmask all these packages, as well as address any of the reverse dependency issues. (You can grab a sneak peak [here if you wish](https://github.com/ezzieyguywuf/gentoo-haskell/tree/quickcheck_fixes_local), I haven't pushed these changes to #1118 yet since it would be confusing with the common commits from this PR. Once this is merged, I'll rebase and #1118 will be more readable and up-to-date)



> Some packages have had their minor versions bumped twice (e.g. `bifunctors-5.5.8` and `bifunctors-5.5.9`). Unless there is a reason for it (and there are reasons to do this, such as minor version bumps depending on different major versions of dependencies, and masking one until such time as it is ready) I will hopefully have a chance to go through post-merge and clean it all out.

I think I took out any intermediate version bumps that weren't necessary, but if I missed any feel free to fix it up.



> I'm hesitant to merge right now due to my restricted/sporadic availability at the moment

😯 I think I can merge it myself now
